### PR TITLE
BP-234 diagram events over time

### PR DESCRIPTION
**Pull request by julianweise**
Medium refactorings were necessary in ProductView to met new requirements of dataloading/architecture.
Please see commit messages for more details.

### CODE REVIEW

**Comment by hpibptlab** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: src/ProductView/Diagram/LineChart.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
+import React, { Component } from 'react';
+import Chart from "chart.js";
+import {argosConfig} from '../../config/argosConfig.js';
+
+class LineChart extends Component {
+    
+    static sortEventsByTime(elementA, elementB, timeStampAttribute) {
+        let datePartsA = elementA[timeStampAttribute].split("T")[0];
+        let datePartsB = elementB[timeStampAttribute].split("T")[0];
+        datePartsA = datePartsA.split("-");
```
Comment (friedow) [inline src/ProductView/Diagram/LineChart.js]: Assigning an array to a string variable makes it harder to read. I suggest using a new const instead.

File: src/ProductView/Diagram/LineChart.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        datePartsB = datePartsB.split("-");
+        const dateA = new Date(datePartsA[0], datePartsA[1], datePartsA[2]);
+        const dateB = new Date(datePartsB[0], datePartsB[1], datePartsB[2]);
+        return dateA - dateB;
+    }
+    
+    buildChartData() {
+        const dataset = [];
+        this.props.eventData.forEach((eventDataContainer) => {
+            dataset.push(this.buildChartDataset(eventDataContainer.eventType, eventDataContainer.events));
```
Comment (friedow) [inline src/ProductView/Diagram/LineChart.js]: Please extract the return value of buildChartDataset to a new variable.

File: src/ProductView/Diagram/LineChart.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        const timeStampAttribute = eventType.timestampAttributeName;
+        let counter = 0;
+        const sortedEvents = events;
+        sortedEvents.sort((elementA, elementB) => {
+            return LineChart.sortEventsByTime(elementA, elementB, timeStampAttribute);
+        });
+        sortedEvents.forEach(function (event) {
+            counter = counter + 1;
+            dataContainer.push({
+                x: event[timeStampAttribute].split("T")[0],
```
Comment (friedow) [inline src/ProductView/Diagram/LineChart.js:32]: The exact same splitting was used in LineChart > sortEventsByTime in line 9. Please extract it to a method.

File: src/ProductView/Diagram/LineChart.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                y: counter
+            });
+        });
+        return({
+            label: eventType.name,
+            fill: true,
+            lineTension: 0.1,
+            backgroundColor: "rgba(0, 78, 100, 0.5)",
```
Comment (friedow) [inline src/ProductView/Diagram/LineChart.js:40]: Extract this to config?

File: src/ProductView/Diagram/LineChart.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            data: dataContainer
+        });
+            
+    }
+
+    componentDidMount() {
+        const charWrapper = this.refs.canvas;
+        if (charWrapper) {
+            const chartContext = charWrapper.getContext('2d');
+            const chart = new Chart(chartContext, {
```
Comment (friedow) [inline src/ProductView/Diagram/LineChart.js]: Extract chart config object to variable.

File: src/ProductView/ProductView.js [hunk 6]
```diff
@@ -16,35 +17,48 @@ class ProductView extends Component {
             product: null,
             eventTypes: null,
             error: null,
-            eventTable: {
-                header: [],
-                events: []
-            }
+            activeEventType: {attributes: []},
+            eventData: [
```
Comment (friedow) [inline src/ProductView/ProductView.js]: No newline here

File: src/ProductView/ProductView.js [hunk 7]
```diff
@@ -95,15 +118,20 @@ class ProductView extends Component {
...
-            this.handleEventData,
-            this.handleError
-        );
+    
+    setActiveEventType(eventType) {
+        this.setState({activeEventType: eventType});
+    }
+    
+    getCurrentEvents() {
+        const eventContainer = this.state.eventData.find((eventContainer) => {
```
Comment (friedow) [inline src/ProductView/ProductView.js:126]: The exact same comparison was used in LineChart > handleEventData in line 37. Please extract it to a method.

File: src/ProductView/ProductView.js [hunk 8]
```diff
@@ -120,17 +148,20 @@ class ProductView extends Component {
                 <div>
                     <Header product={this.state.product}/>
                     <DetailArea product={this.state.product}/>
+                    <LineChart
```
Comment (friedow) [inline src/ProductView/ProductView.js]: No newline here.

File: src/ProductView/Diagram/LineChart.js [hunk 9]
```diff
@@ -0,0 +1,90 @@
+import React, { Component } from 'react';
+import Chart from "chart.js";
+import {argosConfig} from '../../config/argosConfig.js';
+
+class LineChart extends Component {
+    
+    static sortEventsByTime(elementA, elementB, timeStampAttribute) {
+        let datePartsA = elementA[timeStampAttribute].split("T")[0];
```
Comment (friedow) [inline src/ProductView/Diagram/LineChart.js]: Naming suggestion datePartsX => unparsedDateX

**Review by friedow — CHANGES_REQUESTED** [review]

**Review by friedow — APPROVED** [review]

**Pull request merged**